### PR TITLE
Puma

### PR DESCRIPTION
**Pull request by brandur**
Switch off of Unicorn and onto Puma, as it seems to be the web server of
choice these days.

Configuration runs in clustered mode which gives us three separate
worker processes like the Unicorn configuration before it, but each also
runs a collection of threads as well.

@pedro How do you feel about this?

### CODE REVIEW

File: Procfile [hunk 1]
```diff
@@ -1 +1 @@
-web: bundle exec unicorn -p $PORT -c ./config/unicorn.rb
+web: bundle exec puma --quiet --threads 8:32 --port $PORT --preload --workers 3 config.ru
```
Comment (pedro) [inline Procfile]: Can/should we make the number of threads and workers configurable via ENV?

**Comment by pedro** [comment]
Awesome! Definitely down for it :+1:

**Comment by catsby** [comment]
Should probably put min/max threads in `config/puma.rb`, along with `workers`

```
# config/puma.rb
threads Integer(ENV['MIN_THREADS']  || 1), Integer(ENV['MAX_THREADS'] || 16)
workers Integer(ENV['PUMA_WORKERS'] || 3)

port ENV['PORT'] || 3000
environment ENV['RACK_ENV'] || 'development'
preload_app!

Thread.abort_on_exception = true
```

Also, default `8:32` thread count may be high if Sequel defaults to a connection pool of 4, but that largely depends on the workload.

**Comment by brandur** [comment]
@catsby @pedro I'm thinking that since `Procfile` and `config/puma.rb` are part of the template, we could just let people clone it down and adjust the files rather than add in a bunch more config vars. For the most part they'll be extremely invariant. Thoughts?

@catsby +1 moving things into config. Regarding the `8:32`, is there a known optimal config for a Heroku 1X dyno? You're probably right that this is indeed too aggressive.

**Comment by brandur** [comment]
Moved more things to config file and added a more conservative thread count of 1, 16.

**Comment by catsby** [comment]
All of the things I suggested move to the config file come with defaults if the env isn't set, so there shouldn't be a problem there. 

>  is there a known optimal config for a Heroku 1X dyno? You're probably right that this is indeed too aggressive.

This has nothing to do with dyno or even server size. Each thread can request a connection to the db, and Sequel (or ActiveRecord) will give them one from the pool. If the number of active threads exceed database connections, you _may_ have a bad time because threads then have to wait for a connection to open up.  If your queries take a long time, or sufficiently long with sufficient number of waiting threads, you get things like this:

```
ActiveRecord::ConnectionTimeoutError - could not obtain a database connection within 5 seconds. The max pool size is currently 5; consider increasing it` 
```

Not that connection pool must equal thread count, but a high difference could maybe be bad. Again, this depends on your queries.

**Comment by brandur** [comment]
> This has nothing to do with dyno or even server size. Each thread can request a connection to the db, and Sequel (or ActiveRecord) will give them one from the pool. If the number of active threads exceed database connections, you may have a bad time because threads then have to wait for a connection to open up. If your queries take a long time, or sufficiently long with sufficient number of waiting threads, you get things like this:

Oh yes totally get it! I guess that question was unrelated to the question of thread pool size. I figure that if we optimize thread/worker count for a dyno, we can then bring up the size of the thread pool to reasonably pace it.

**Comment by catsby** [comment]
@schneems may have info for tuning to a 1x dyno, but it may also just be dependent on the application. 

I think the 3 workers, 1-16 threads is probably a reasonable default

**Comment by pedro** [comment]
> @catsby @pedro I'm thinking that since Procfile and config/puma.rb are part of the template, we could just let people clone it down and adjust the files rather than add in a bunch more config vars. For the most part they'll be extremely invariant. Thoughts?

I think it deserves a config var because it's a value people tend to mess around when running load tests or moving across dyno sizes. If we establish that pattern early (and have nice defaults like you guys mentioned) then people won't have to push again to change. Thoughts?

**Comment by brandur** [comment]
@pedro Cool, I don't feel too strongly about it. Added. I think I'm going to try for a @mfine-style config solution on top of this to reduce the verbosity of these checks.

Pulling this in for now. Thanks for taking a look guys!

**Pull request merged**